crates/sui-framework/sources/versioned_type.move

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module sui::versioned_type {
+    use sui::object::{UID, ID};
+    use sui::tx_context::TxContext;
+    use sui::object;
+    use sui::dynamic_field;
+
+    const EInvalidUpgrade: u64 = 0;
+
+    /// A wrapper type that supports versioning of the inner type.
+    /// The inner type is a dynamic field of the Versioned object, and is keyed using version.
+    struct Versioned has key, store {
+        id: UID,
+        version: u64,
+    }
+
+    /// Represents a hot potato object generated when we take out the dynamic field.
+    /// This is to make sure that we always put a new value back.
+    struct VersionChangeCap {
+        versioned_id: ID,
+        old_version: u64,
+    }
+
+    public fun create<T: store>(init_version: u64, init_value: T, ctx: &mut TxContext): Versioned {
```

Comment:
Can we add doc comments to the public functions?  Mainly because I think this library will be more broadly useful for people building on Sui!